### PR TITLE
[14.0][MIG] sale_invoice_policy

### PR DESCRIPTION
**Pull request by enriquemartin**
Migration + some changes:
- Simplified field management in `res.config.settings`: removed duplicate fields to use the ones added by sale (`default_invoice_policy`).
- Moved `post-init-hook` to `pre-init-hook` in order to keep the default method of the `default_invoice_policy` field.
- Same literals for `invoice_policy` field as in sale
- Updated pot file + Spanish translation

### CODE REVIEW

**Comment by damdam-s** [comment]
travis error is not related to this module.

However, can you please create 1 commit containing the "normal" migration to v14 and another one in which you do the evolutions/changes please?

Changes will be easier to review w/ this process.
thanks a lot in advance

**Comment by enriquemartin** [comment]
That was my initial idea: do a standard migration and then propose changes. But in the process of reviewing the code,  I encountered some issues that did not work without changes. E.g. the default value of `default_invoice_policy` field is not working on a first installation, because of that the `post_init_hook` has to be changed to `pre_init_hook`. 
I will separate the rest of the changes in a separate commit.
Thanks!

**Comment by enriquemartin** [comment]
Done @damdam-s two separate commits to facilitate review!

**Comment by enriquemartin** [comment]
any news on this @damdam-s? thanks in advance!

**Review by jonerikceberio — COMMENTED** [review]
LGTM

**Review by jonerikceberio — APPROVED** [review]

File: sale_invoice_policy/models/product_template.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        Apply the invoice_policy given by context (if exist) otherwise use the
+        default invoice policy given by the field with this same name.
+        If the product is type = 'service', we don't have to apply the invoice
+        policy given by the context.
+        :return:
+        """
+        invoice_policy = self.env.context.get("invoice_policy")
+        for tmpl in self:
+            if tmpl.type != "service" and invoice_policy:
```
Comment (flotho) [inline sale_invoice_policy/models/product_template.py:48]: Could you please precise why you make a difference with service products.
either 'service' and 'product' could be invoiced based on ordered or delivered quantity.

File: sale_invoice_policy/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            ("order", "Invoice what is ordered"),
+            ("delivery", "Invoice what is delivered"),
+        ],
+        string="Invoicing Policy",
+        readonly=True,
+        states={"draft": [("readonly", False)], "sent": [("readonly", False)]},
+        help="Ordered Quantity: Invoice quantities ordered by the customer.\n"
+        "Delivered Quantity: Invoice quantities delivered to the customer.",
+    )
+    invoice_policy_required = fields.Boolean(default=False, readonly=True)
```
Comment (flotho) [inline sale_invoice_policy/models/sale_order.py]: sorry if I didn't read correctly but could you precise where this field is used, I can see a view but no python part using this variable.

File: sale_invoice_policy/models/sale_order_line.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        "qty_invoiced",
+        "qty_delivered",
+        "product_uom_qty",
+        "order_id.state",
+        "order_id.invoice_policy",
+    )
+    def _get_to_invoice_qty(self):
+        invoice_policies = set(self.mapped("order_id.invoice_policy"))
+        line_by_id = {sol.id: sol for sol in self}
+        done_lines = self.env["sale.order.line"].browse()
```
Comment (flotho) [inline sale_invoice_policy/models/sale_order_line.py]: won't it return all the order.line of the database?

**Comment by enriquemartin** [comment]
@i-vyshnevska any chance to review this? Thanks in advance!

File: sale_invoice_policy/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    "version": "14.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["sale", "sale_stock"],
+    "data": [
+        "data/sale_data.xml",
+        "wizards/res_config_view.xml",
+        "views/product_template_view.xml",
+        "views/sale_view.xml",
+    ],
+    "pre_init_hook": "pre_init_hook",
```
Comment (i-vyshnevska) [inline sale_invoice_policy/__manifest__.py:18]: should stay in post_init_hook as in pre-state "invoice_policy" field will not de recomputed with code presented in this addon

File: sale_invoice_policy/models/sale_order_line.py (hunk 3 above)
Comment (i-vyshnevska) [reply sale_invoice_policy/models/sale_order_line.py]: no that will be empty model this is equal to 
`done_lines = self.env["sale.order.line"]`

File: sale_invoice_policy/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        # were not in so_lines
+        super(SaleOrderLine, self - done_lines)._get_to_invoice_qty()
+        return True
+
+    @api.depends(
+        "state",
+        "product_uom_qty",
+        "qty_delivered",
+        "qty_to_invoice",
+        "qty_invoiced",
```
Comment (i-vyshnevska) [inline sale_invoice_policy/models/sale_order_line.py]:
```suggestion
```
dependency on these fields injected in parent method so we don't need to have it here

File: sale_invoice_policy/models/product_template.py (hunk 1 above)
Comment (enriquemartin) [reply sale_invoice_policy/models/product_template.py:48]: You are right, we no longer make a distinction.

File: sale_invoice_policy/models/sale_order.py (hunk 2 above)
Comment (enriquemartin) [reply sale_invoice_policy/models/sale_order.py]: Field removed. Indeed it was not used after code simplification.

File: sale_invoice_policy/models/sale_order_line.py (hunk 3 above)
Comment (enriquemartin) [reply sale_invoice_policy/models/sale_order_line.py]: Modified to simplify understanding

File: sale_invoice_policy/models/sale_order_line.py (hunk 5 above)
Comment (enriquemartin) [reply sale_invoice_policy/models/sale_order_line.py]: Removed!

File: sale_invoice_policy/__manifest__.py (hunk 4 above)
Comment (enriquemartin) [reply sale_invoice_policy/__manifest__.py:18]: I do not understand the problem. The _pre_init_hook()_ just creates in SQL the new _default_invoice_policy_ column and initializes it with the same value as _invoice_policy_.
It does not need any python code at all.

Comment (i-vyshnevska) [reply sale_invoice_policy/__manifest__.py:18]: nevermind, you are right

**Review by i-vyshnevska — APPROVED** [review]

**Review by damdam-s — APPROVED** [review]
thanks

File: sale_invoice_policy/models/product_template.py (hunk 1 above)
Comment (enriquemartin) [reply sale_invoice_policy/models/product_template.py:48]: I was not right. We need to make a distinction, for example, on orders with storable product and shipping charges: if we mark the billing policy as delivered quantities on the order, shipping charges will never be billed (unless the delivered quantity of this service is forced).
I reverse this change, and leave a comment in the code.

**Review by flotho — APPROVED** [review]
thanks for your explanations

**Review by flotho — APPROVED** [review]

**Comment by enriquemartin** [comment]
In order to be compatible with other addons as `sale_automatic_workflow`, the module cannot set the default value for _invoice_policy_ on `sale.order`, so I have deleted the `default_get` method.
An empty value on field _invoice_policy_ on sale order means that the used value will be the one in each product.

File: sale_invoice_policy/models/sale_order.py [hunk 6]
```diff
@@ -1,43 +1,22 @@
...
         readonly=True,
         states={"draft": [("readonly", False)], "sent": [("readonly", False)]},
-        help="Ordered Quantity: Invoice based on the quantity the customer "
-        "ordered.\n"
-        "Delivered Quantity: Invoiced based on the quantity the vendor "
-        "delivered (time or deliveries).",
+        help="Ordered Quantity: Invoice quantities ordered by the customer.\n"
+        "Delivered Quantity: Invoice quantities delivered to the customer.",
     )
-    invoice_policy_required = fields.Boolean(
```
Comment (rousseldenis) [inline sale_invoice_policy/models/sale_order.py]: This module changes the standard behaviour. It's better to keep it optional.

Comment (enriquemartin) [reply sale_invoice_policy/models/sale_order.py]: I don't understand what you are talking about: are you meaning the deletion of the `invoice_policy_required` field is not "standard"?
As I said the option to make required `invoice_policy` on sale.order makes this module not compatible with others like `sale_automatic_workflow`.

Comment (bunhin) [reply sale_invoice_policy/models/sale_order.py]: It seems the comment from @rousseldenis has made this PR review process not active, while asked question not yet replied

**Comment by elvise** [comment]
Hi @enriquemartin and thanks for your porting ! 
Can you rebase ?
Thanks in advance :)

**Review by ivantodorovich — APPROVED** [review]

**Comment by enriquemartin** [comment]
@dreispt why the needs fixing tag?

**Comment by dreispt** [comment]
@enriquemartin That looks like a mistake. I guess this is ready to merge!

**Pull request merged**